src/PerformanceSensitiveAnalyzers/Core/ExplicitAllocationAnalyzer.cs

```diff
@@ -0,0 +1,109 @@
+
+        internal static DiagnosticDescriptor ArrayCreationRule = new DiagnosticDescriptor(
+            ArrayCreationRuleId,
+            s_localizableArrayCreationRuleTitleAndMessage,
+            s_localizableArrayCreationRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        internal static DiagnosticDescriptor ObjectCreationRule = new DiagnosticDescriptor(
+            ObjectCreationRuleId,
+            s_localizableObjectCreationRuleTitleAndMessage,
+            s_localizableObjectCreationRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        internal static DiagnosticDescriptor AnonymousObjectCreationRule = new DiagnosticDescriptor(
+            AnonymousObjectCreationRuleId,
+            s_localizablAnonymousObjectCreationRuleTitleAndMessage,
+            s_localizablAnonymousObjectCreationRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true,
+            helpLinkUri: "http://msdn.microsoft.com/en-us/library/bb397696.aspx");
+
+        internal static DiagnosticDescriptor LetCauseRule = new DiagnosticDescriptor(
+            LetCauseRuleId,
+            s_localizableLetCauseRuleTitleAndMessage,
+            s_localizableLetCauseRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        private static readonly object[] EmptyMessageArgs = Array.Empty<object>();
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            ArrayCreationRule,
+            ObjectCreationRule,
+            AnonymousObjectCreationRule,
+            LetCauseRule);
+
+        protected override ImmutableArray<OperationKind> Operations => ImmutableArray.Create(
+            OperationKind.ArrayCreation,
+            OperationKind.ObjectCreation,
+            OperationKind.AnonymousObjectCreation);
+
+        protected override void AnalyzeNode(OperationAnalysisContext context, in PerformanceSensitiveInfo info)
+        {
+            if (context.Operation is IArrayCreationOperation)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(ArrayCreationRule, context.Operation.Syntax.GetLocation(), EmptyMessageArgs));
+                return;
+            }
+
+            if (context.Operation is IObjectCreationOperation)
+            {
+                if (context.Operation.Type.IsReferenceType)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(ObjectCreationRule, context.Operation.Syntax.GetLocation(), EmptyMessageArgs));
+                    return;
+                }
+
+                if (context.Operation.Parent.GetType().Name == "LazyConversionOperation")
```

Comment:
❓ Why do we need this type check?

Reply:
Good question, we can just as well use `if (context.Operation.Parent is IConversionOperation)`

The actual check relates to the question I posed whether this scenario really belongs here. It catches the case:
```using Roslyn.Utilities;

public struct S { }

public class MyClass
{
    [PerformanceSensitive(""uri"")]
    public void Foo() 
    {
        object box = new S();
    }
}`

Reply:
Checking against the interface would be much better I believe.